File: implement-shell-tools/wc/my-wc.js
#!/usr/bin/env node

const fs = require("node:fs");

function countFile(filePath) {
  const stats = fs.statSync(filePath);

  if (!stats.isFile()) {
    console.warn(`${filePath} is not a file, skipping`);
    return null;
  }
  const content = fs.readFileSync(filePath, "utf8");

  const lines = content.split("\n").length - 1;
  const words = content.trim() ? content.trim().split(/\s+/).length : 0;
  const chars = Buffer.byteLength(content, "utf8");

  return { lines, words, chars };
}

function main() {
  const args = process.argv.slice(2);

  let flags = [];
  let files = [];

  // 1 Parse args
  for (const arg of args) {
    if (arg === "-l" || arg === "-w" || arg === "-c") {
      flags.push(arg);
    } else {
      files.push(arg);
    }
  }

  //2 Helper function that decides what to print

  function formatOutput(counts, files) {
    const parts = [];

    // if no flags show everything
    const showAll = flags.length === 0;

    if (showAll || flags.includes("-l"))parts.push(counts.lines);
    if (showAll || flags.includes("-w"))parts.push(counts.words);
    if (showAll || flags.includes("-c"))parts.push(counts.chars);

    parts.push(files);

    return parts.join(" ");
  }
  // 3 totals
  let totalLines = 0;
  let totalWords = 0;
  let totalChars = 0;


  // 4 per-file output
  for (const file of files) {
    const counts = countFile(file);
    if (!counts) continue;

    totalLines += counts.lines;
    totalWords += counts.words;
    totalChars += counts.chars;

    console.log(formatOutput(counts, files));
  }
  // 5 Total output (only if multiple files)
  if (files.length > 1) {
    const totalCounts = {
      lines: totalLines,
      words: totalWords,
      chars: totalChars,
    };
    console.log(formatOutput(totalCounts,"total"));
  }
}

main();

File: implement-shell-tools/wc/my-wc.py
#!/usr/bin/env python3

import sys


def count_file(path):
    with open(path, "rb") as f:
        content = f.read()

    byte_count = len(content)
    text = content.decode("utf-8", errors="ignore")

    line_count = text.count("\n")
    word_count = len(text.split())

    return line_count, word_count, byte_count



def main():
    args = sys.argv[1:]

    show_l = False
    show_w = False
    show_c = False

    paths = []
    
    
    # parse args
    for a in args:
        if a == "-l":
            show_l = True
        elif a == "-w":
            show_w = True
        elif a == "-c":
            show_c = True
        else:
            paths.append(a)

    # default: show all
    if not (show_l or show_w or show_c):
        show_l = show_w = show_c = True

    files = paths

    total_l = 0
    total_w = 0
    total_c = 0
    
    results = []
    had_error = False
    
    for file in files:
        try:
            l, w, c = count_file(file)
        except FileNotFoundError:
            print(
                f"wc: {file}: No such file or directory",
                file=sys.stderr,
            )
            had_error = True
            continue
        
        total_l += l
        total_w += w
        total_c += c
        
        results.append((l,w,c, file))
        
     # print per-file results (GNU-aligned formatting)
    for l, w, c, file in results:

        parts = []
        if show_l:
            parts.append(f"{l:3}")
        if show_w:
            parts.append(f"{w:4}")
        if show_c:
            parts.append(f"{c:4}")


        print("".join(parts) + " " + file)
        
    # print total if multiple files
    if len(results) > 1:
        parts = []

        if show_l:
            parts.append(f"{total_l:3}")
        if show_w:
            parts.append(f"{total_w:4}")
        if show_c:
            parts.append(f"{total_c:4}")

        
        print("".join(parts) + " total")
        
    if  had_error:
        sys.exit(1)   


if __name__ == "__main__":
    main()